share/ci/pypicongpu_generator.py

```diff
@@ -0,0 +1,319 @@
+import requests
+import sys
+import yaml
+import re
+from typing import List, Dict, Callable
+import pkg_resources
+
+"""
+This file is part of the PIConGPU.
+Copyright 2023 PIConGPU contributors
+Authors: Simeon Ehrig
+License: GPLv3+
+"""
+
+"""@file Generate different CI test jobs for different Python version and
+         depending on a requirement.txt.
+
+Prints yaml code for a GitLab CI child pipeline to stout. The test parameter
+are split in two kinds of inputs. The Python versions to test are defined in
+the script, also the names of dependencies to test and it's test strategy. The
+version range of the dependencies to test are defined in a requirements.txt.
+The path of the requirements.txt is set via first application argument.
+
+First, the script reads the requirements.txt. If a dependency is marked to test
+in the script, it calculates the test versions. Therefore it downloads all
+available versions from pypi.org for each package. Afterwards it filters the
+versions via a filter strategy. For example, take all release versions or take
+each latest major version. Than the script removes all versions, which are not
+supported, defined in the requirements.txt. The result a complete list of all
+Python and dependencies versions to test.
+
+In the second part, the script creates the full combination matrix for all test
+versions and creates a CI job for each combination. Each job is printed on
+stdout.
+
+The number of combinations depends on:
+- number of supported Python version
+- number of dependencies to test
+- the test strategy of each dependencies to test
+- versions restrictions in the requirements.txt
+- releases of the dependencies
+
+@param First application argument: Path to the requirements.txt
+"""
+
+
+# caches the parsed dependencies of the requirements.txt here
+req_versions: Dict[str, pkg_resources.Requirement] = {}
+
+
+def cs(text: str, color: str) -> str:
+    """Print the text in a different color on the command line. The text after
+       the function has the default color of the command line.
+
+    Parameters
+    ----------
+        text (str): text to be colored
+        color (str): Name of the color. If wrong color or empty, use default
+                    color of the command line.
+
+    Returns
+    -------
+        str: text with bash pre and post string for coloring
+    """
+
+    if color is None:
+        return text
+
+    output = ""
+    if color == "Red":
+        output += "\033[0;31m"
+    elif color == "Green":
+        output += "\033[0;32m"
+    elif color == "Yellow":
+        output += "\033[1;33m"
+
+    return output + text + "\033[0m"
+
+
+def get_all_pypi_versions(package_name: str) -> List[str]:
+    """Returns all release versions of a package registered on pypi.org
+
+    Parameters
+    ----------
+        package_name (str): Name of the searched package.
+
+    Returns
+    -------
+        List[str]: List of release versions.
+    """
```

Comment:
is there a reason why you are sometimes using `@param` and some times not?

Reply:
I copied the doc string style of the functions from other python scripts. Therefore it is inconsistent. I can fix it.